File: mobile/lib/data/repositories/weight_estimation_repository_impl.dart
/// Repository Implementation: WeightEstimationRepositoryImpl
///
/// Implementación del repositorio de estimación de peso.
/// Single Responsibility: Coordinar TFLite y almacenamiento local.
///
/// Data Layer - Clean Architecture
library;

import 'package:dartz/dartz.dart';
import 'package:flutter/foundation.dart';

import 'dart:io';

import '../../core/constants/breeds.dart';
import '../../core/errors/exceptions.dart';
import '../../core/errors/failures.dart';
import '../../domain/entities/weight_estimation.dart';
import '../../domain/repositories/weight_estimation_repository.dart';
import '../datasources/ml_remote_datasource.dart';
import '../datasources/tflite_datasource.dart';
import '../datasources/weight_estimation_local_datasource.dart';
import '../models/weight_estimation_model.dart';

/// Implementación del repositorio de estimación de peso
class WeightEstimationRepositoryImpl implements WeightEstimationRepository {
  final TFLiteDataSource tfliteDataSource;
  final WeightEstimationLocalDataSource localDataSource;
  final MLRemoteDataSource? mlRemoteDataSource; // Opcional para fallback

  WeightEstimationRepositoryImpl({
    required this.tfliteDataSource,
    required this.localDataSource,
    this.mlRemoteDataSource,
  });

  @override
  Future<Either<Failure, WeightEstimation>> estimateWeight({
    required String imagePath,
    required BreedType breed,
    String? cattleId,
  }) async {
    // Intentar primero con el servidor (si está disponible)
    if (mlRemoteDataSource != null) {
      try {
        final imageFile = File(imagePath);
        if (await imageFile.exists()) {
          final prediction = await mlRemoteDataSource!.predictWeight(
            imageFile: imageFile,
            breed: breed.value,
            animalId: cattleId,
          );

          // Convertir respuesta del servidor a WeightEstimation entity
          final estimation = WeightEstimation(
            id: prediction.id,
            cattleId: prediction.animalId,
            breed: breed,
            estimatedWeight: prediction.estimatedWeightKg,
            confidenceScore: prediction.confidence,
            frameImagePath: imagePath,
            timestamp: prediction.timestamp,
            method: EstimationMethod.tflite, // El servidor usa TFLite
            modelVersion: prediction.mlModelVersion,
            processingTimeMs: prediction.processingTimeMs,
          );

          return Right(estimation);
        }
      } on NetworkException catch (e) {
        // Si no hay conexión, usar TFLite local como fallback
        debugPrint('⚠️ Sin conexión, usando TFLite local: ${e.message}');
      } on ServerException catch (e) {
        // Si el servidor falla, usar TFLite local como fallback
        debugPrint('⚠️ Error del servidor, usando TFLite local: ${e.message}');
      } catch (e) {
        // Cualquier otro error, usar TFLite local como fallback
        debugPrint('⚠️ Error con servidor, usando TFLite local: $e');
      }
    }

    // Fallback: Ejecutar inferencia con TFLite local
    try {
      final estimation = await tfliteDataSource.runInference(
        imagePath: imagePath,
        breed: breed,
        cattleId: cattleId,
      );

      // Validar que el tiempo de procesamiento sea <3s
      if (estimation.processingTimeMs >= 3000) {
        debugPrint(
          '⚠️ Procesamiento lento: ${estimation.processingTimeMs}ms (objetivo: <3000ms)',
        );
      }

      return Right(estimation);
    } on ModelException catch (e) {
      return Left(InferenceFailure(message: e.message));
    } on StorageException catch (e) {
      return Left(StorageFailure(message: e.message));
    } catch (e) {
      return Left(UnknownFailure(message: 'Error en estimación: $e'));
    }
  }

  @override
  Future<Either<Failure, void>> saveEstimation(
    WeightEstimation estimation,
  ) async {
    try {
      final estimationModel = WeightEstimationModel.fromEntity(estimation);
      await localDataSource.saveEstimation(estimationModel);
      return const Right(null);
    } on DatabaseException catch (e) {
      return Left(DatabaseFailure(message: e.message));
    } catch (e) {
      return Left(UnknownFailure(message: 'Error al guardar: $e'));
    }
  }

  @override
  Future<Either<Failure, List<WeightEstimation>>> getAllEstimations() async {
    try {
      final estimations = await localDataSource.getAllEstimations();
      return Right(estimations);
    } on DatabaseException catch (e) {
      return Left(DatabaseFailure(message: e.message));
    } catch (e) {
      return Left(UnknownFailure(message: 'Error al obtener estimaciones: $e'));
    }
  }

  @override
  Future<Either<Failure, List<WeightEstimation>>> getEstimationsByCattle(
    String cattleId,
  ) async {
    try {
      final estimations = await localDataSource.getEstimationsByCattle(
        cattleId,
      );
      return Right(estimations);
    } on DatabaseException catch (e) {
      return Left(DatabaseFailure(message: e.message));
    } catch (e) {
      return Left(
        UnknownFailure(message: 'Error al obtener estimaciones del animal: $e'),
      );
    }
  }

  @override
  Future<Either<Failure, WeightEstimation?>> getLastEstimation(
    String cattleId,
  ) async {
    try {
      final estimation = await localDataSource.getLastEstimation(cattleId);
      return Right(estimation);
    } on DatabaseException catch (e) {
      return Left(DatabaseFailure(message: e.message));
    } catch (e) {
      return Left(
        UnknownFailure(message: 'Error al obtener última estimación: $e'),
      );
    }
  }

  @override
  Future<Either<Failure, void>> deleteEstimation(String estimationId) async {
    try {
      await localDataSource.deleteEstimation(estimationId);
      return const Right(null);
    } on DatabaseException catch (e) {
      return Left(DatabaseFailure(message: e.message));
    } catch (e) {
      return Left(UnknownFailure(message: 'Error al eliminar estimación: $e'));
    }
  }

  @override
  Future<Either<Failure, bool>> areModelsLoaded() async {
    try {
      final loaded = await tfliteDataSource.areModelsLoaded();
      return Right(loaded);
    } on ModelException catch (e) {
      return Left(ModelLoadFailure(message: e.message));
    } catch (e) {
      return Left(UnknownFailure(message: 'Error al verificar modelos: $e'));
    }
  }

  @override
  Future<Either<Failure, void>> loadModels([List<BreedType>? breeds]) async {
    try {
      // Si no se especifican razas, cargar todas las 7
      final breedsList = breeds ?? BreedType.values;

      await tfliteDataSource.loadModels(breedsList);

      return const Right(null);
    } on ModelException catch (e) {
      return Left(ModelLoadFailure(message: e.message));
    } catch (e) {
      return Left(UnknownFailure(message: 'Error al cargar modelos: $e'));
    }
  }
}

File: backend/app/data/repositories/weight_estimation_repository_impl.py
"""
Weight Estimation Repository Implementation - Data Layer
Implementación del repositorio usando Beanie ODM
"""

from uuid import UUID

from ...domain.entities.weight_estimation import WeightEstimation
from ...domain.repositories.weight_estimation_repository import (
    WeightEstimationRepository,
)
from ..models.weight_estimation_model import WeightEstimationModel


class WeightEstimationRepositoryImpl(WeightEstimationRepository):
    """
    Implementación del repositorio de estimaciones usando Beanie.

    Single Responsibility: Persistencia de estimaciones en MongoDB.
    """

    async def find_by_id(self, estimation_id: UUID) -> WeightEstimation | None:
        """Busca una estimación por ID."""
        model = await WeightEstimationModel.get(estimation_id)
        if model is None:
            return None
        return self._to_entity(model)

    async def create(self, estimation: WeightEstimation) -> WeightEstimation:
        """Crea una nueva estimación."""
        model = self._to_model(estimation)
        await model.insert()
        return self._to_entity(model)

    async def update(self, estimation: WeightEstimation) -> WeightEstimation:
        """Actualiza una estimación existente."""
        model = await WeightEstimationModel.get(estimation.id)
        if model is None:
            raise ValueError(f"Estimación {estimation.id} no encontrada")

        # Actualizar campos
        model.animal_id = estimation.animal_id
        model.breed = estimation.breed
        model.estimated_weight_kg = estimation.estimated_weight_kg
        model.confidence = estimation.confidence
        model.method = estimation.method
        model.model_version = estimation.model_version
        model.processing_time_ms = estimation.processing_time_ms
        model.frame_image_path = estimation.frame_image_path
        model.latitude = estimation.latitude
        model.longitude = estimation.longitude
        model.timestamp = estimation.timestamp
        model.device_id = estimation.device_id
        model.synced_at = estimation.synced_at

        await model.save()
        return self._to_entity(model)

    async def count(self) -> int:
        """Retorna el conteo total de estimaciones."""
        return await WeightEstimationModel.count()

    def _to_entity(self, model: WeightEstimationModel) -> WeightEstimation:
        """Convierte Model a Entity."""
        return WeightEstimation(
            id=model.id,
            animal_id=model.animal_id,
            breed=model.breed,
            estimated_weight_kg=model.estimated_weight_kg,
            confidence=model.confidence,
            method=model.method,
            model_version=model.model_version,
            processing_time_ms=model.processing_time_ms,
            frame_image_path=model.frame_image_path,
            latitude=model.latitude,
            longitude=model.longitude,
            timestamp=model.timestamp,
            device_id=model.device_id,
            synced_at=model.synced_at,
        )

    def _to_model(self, entity: WeightEstimation) -> WeightEstimationModel:
        """Convierte Entity a Model."""
        return WeightEstimationModel(
            id=entity.id,
            animal_id=entity.animal_id,
            breed=entity.breed,
            estimated_weight_kg=entity.estimated_weight_kg,
            confidence=entity.confidence,
            method=entity.method,
            model_version=entity.model_version,
            processing_time_ms=entity.processing_time_ms,
            frame_image_path=entity.frame_image_path,
            latitude=entity.latitude,
            longitude=entity.longitude,
            timestamp=entity.timestamp,
            device_id=entity.device_id,
            synced_at=entity.synced_at,
        )
